### ISSUE TITLE
vmtoolsd --cmd sometimes hangs after repeated calls

### ISSUE THREAD

**Issue by yvespp** [vmware/open-vm-tools#377]
I noticed that running `vmtoolsd --cmd 'info-get guestinfo.vendordata'` in a loop sometimes hangs for about two seconds.
```
root@create-test:/var/log# while true; do time /usr/bin/vmtoolsd --cmd 'info-get 
guestinfo.vendordata' done
No value found

real    0m0.009s
user    0m0.008s
sys     0m0.000s
No value found

real    0m2.019s
user    0m0.000s
sys     0m0.008s
No value found

real    0m0.008s
user    0m0.008s
sys     0m0.000s
No value found

real    0m2.040s
user    0m0.004s
sys     0m0.004s
No value found

real    0m0.012s
user    0m0.006s
sys     0m0.006s
No value found

real    0m0.012s
user    0m0.011s
sys     0m0.000s
No value found
```
If I add a sleep >= 50 ms in the loop the command doesn't hang.
`vmware-rpctool 'info-get guestinfo.vendordata'` doesn't show this behaviour and it also seems to be faster in general:
```
root@yp-test-master-4fd795584e81:~# while true; do time vmware-rpctool 'info-get guestinfo.vendordata';  done;
No value found

real    0m0.002s
user    0m0.002s
sys     0m0.000s
No value found

real    0m0.001s
user    0m0.000s
sys     0m0.001s
No value found

real    0m0.001s
user    0m0.001s
sys     0m0.000s
No value found

real    0m0.001s
user    0m0.001s
sys     0m0.000s
No value found

real    0m0.001s
user    0m0.001s
sys     0m0.000s
No value found

real    0m0.002s
user    0m0.001s
sys     0m0.000s
No value found

real    0m0.001s
user    0m0.001s
sys     0m0.000s
No value found

real    0m0.001s
user    0m0.001s
sys     0m0.000s
No value found
```

Originally I opened an issue in https://github.com/vmware/cloud-init-vmware-guestinfo/ as it uses `vmtoolsd` to get the cloud-init data and this command slows down the system boot.
For details see issue: https://github.com/vmware/cloud-init-vmware-guestinfo/issues/18

Why is one slower than the other?
Is it save to use `vmware-rpctool` instead of `vmtoolsd` to get the cloud-init data in cloud-init-vmware-guestinfo? Is `vmware-rpctool` included in all distros that include `vmtoolsd`?

**Comment by oliverkurth** [comment]
Thanks for reporting. We are able to reproduce this issue, but currently we do not have a solution.

The difference between `vmtoolsd` and `vmware-rpctool` is because the former uses vsock and the latter the backdoor.  `vmware-rpctool` should be part of any distribution as part of the `open-vm-tools` package.

**Comment by ravindravmw** [comment]
> If I add a sleep >= 50 ms in the loop the command doesn't hang.

@yvespp unfortunately, adding a small sleep between 2 invocations (as you already tried) is the only workaround at this point until the issue gets fixed in open-vm-tools and our products.

> Why is one slower than the other?

As you have noticed vSocket is slightly slower than Backdoor because of the way it is handled by the hypervisor. However, this is likely to change in future.

> Is it save to use vmware-rpctool instead of vmtoolsd to get the cloud-init data in cloud-init-vmware-guestinfo?

Please be aware that privileged commands (commands that need caller to have root privilege) can be invoked over vSocket only. For these reasons, we don't recommend using vmware-rpctool in general.  However, you should be able to use vmware-rpctool in the interim because the command you are invoking does not require privileged access.

> Is vmware-rpctool included in all distros that include vmtoolsd?

Oliver answered already.

**Comment by stanguturi** [comment]
> > If I add a sleep >= 50 ms in the loop the command doesn't hang.
> 
> @yvespp unfortunately, adding a small sleep between 2 invocations (as you already tried) is the only workaround at this point until the issue gets fixed in open-vm-tools and our products.

There are two parts (open-vm-tools and hypervisor) to this fix.

The fix in open-vm-tools has been made in bf9de87b3cd4d25e6ca23db63c01f353acf40419 
The fix in VMware hypervisor has been made in the latest ESXi release. (starting from ESXi 7.0 U1). Check https://kb.vmware.com/s/article/2143832 for information about the various ESXi releases and build details.

So, if you take *both* the latest ESXi and latest Open-vm-tools, the issue should not be there.
